File: src/main/java/net/andrecarbajal/naviMusic/audio/TrackScheduler.java
package net.andrecarbajal.naviMusic.audio;

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer;
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter;
import com.sedmelluq.discord.lavaplayer.track.AudioTrack;
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason;
import lombok.Getter;
import lombok.RequiredArgsConstructor;
import lombok.Setter;
import lombok.extern.slf4j.Slf4j;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.entities.Member;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.BlockingQueue;
import java.util.concurrent.Executors;
import java.util.concurrent.LinkedBlockingQueue;
import java.util.concurrent.ScheduledExecutorService;
import java.util.concurrent.ScheduledFuture;
import java.util.concurrent.TimeUnit;

@Slf4j
@RequiredArgsConstructor
public class TrackScheduler extends AudioEventAdapter {
    private final AudioPlayer player;
    private final Guild guild;
    @Getter
    private final BlockingQueue<AudioTrack> queue = new LinkedBlockingQueue<>();
    @Getter
    @Setter
    private boolean repeating = false;

    private final ScheduledExecutorService executor = Executors.newSingleThreadScheduledExecutor();
    private ScheduledFuture<?> disconnectTask;

    public void queue(AudioTrack track) {
        log.info("Adding {} in the queue", track.getInfo().title);
        cancelDisconnectTimer();
        if (player.startTrack(track, true)) {
            logTackStarted(track);
        } else {
            if (!queue.offer(track)) throw new RuntimeException("Item was not added to the queue");
        }
    }

    public void nextTrack() {
        AudioTrack track = queue.poll();
        if (track != null) {
            logTackStarted(track);
        }
        player.startTrack(track, false);
    }

    public void skipTrack(int position) {
        if (position < 1) return;

        if (position == 1) {
            nextTrack();
            return;
        }

        List<AudioTrack> tracks = new ArrayList<>(queue);
        tracks.remove(position - 1);
        queue.clear();
        queue.addAll(tracks);
    }

    public void clear() {
        queue.clear();
        nextTrack();
        checkDisconnect();
    }

    public int getQueueSize() {
        return queue.size();
    }

    public void shuffle() {
        List<AudioTrack> tracks = new ArrayList<>(queue);
        queue.clear();
        Collections.shuffle(tracks);
        tracks.forEach(this::queue);
    }

    public void onTrackEnd(AudioPlayer player, AudioTrack track, AudioTrackEndReason endReason) {
        if (endReason.mayStartNext) {
            if (repeating) {
                player.startTrack(track.makeClone(), false);
                checkDisconnect();
                return;
            }
            nextTrack();
            checkDisconnect();
            return;
        }

        if (endReason == AudioTrackEndReason.LOAD_FAILED) {
            log.error("Error playing {} ({})", track.getInfo().title, track.getInfo().uri);
            nextTrack();
            checkDisconnect();
        }

        checkDisconnect();
    }

    @Override
    public void onTrackStart(AudioPlayer player, AudioTrack track) {
        cancelDisconnectTimer();
    }

    private void checkDisconnect() {
        if (queue.isEmpty() && player.getPlayingTrack() == null) {
            scheduleDisconnect();
        }
    }

    private void scheduleDisconnect() {
        if (disconnectTask != null) {
            disconnectTask.cancel(false);
        }
        disconnectTask = executor.schedule(() -> guild.getAudioManager().closeAudioConnection(), 3, TimeUnit.MINUTES);
    }

    private void cancelDisconnectTimer() {
        if (disconnectTask != null) {
            disconnectTask.cancel(false);
            disconnectTask = null;
        }
    }

    private void logTackStarted(AudioTrack track) {
        Member member = track.getUserData(Member.class);

        if (member != null) {
            log.info("Playing {} in {} requested by {}", track.getInfo().title, member.getGuild().getName(), member.getEffectiveName());
        } else {
            log.info("Playing {}", track.getInfo().title);
        }
    }
}

File: src/main/kotlin/net/andrecarbajal/naviMusic/audio/TrackScheduler.kt
package net.andrecarbajal.naviMusic.audio

import com.sedmelluq.discord.lavaplayer.player.AudioPlayer
import com.sedmelluq.discord.lavaplayer.player.event.AudioEventAdapter
import com.sedmelluq.discord.lavaplayer.track.AudioTrack
import com.sedmelluq.discord.lavaplayer.track.AudioTrackEndReason
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.entities.Member
import org.slf4j.LoggerFactory
import java.util.concurrent.*

class TrackScheduler(private val player: AudioPlayer, private val guild: Guild) : AudioEventAdapter() {

    private val log = LoggerFactory.getLogger(TrackScheduler::class.java)

    val queue: BlockingQueue<AudioTrack> = LinkedBlockingQueue()
    var isRepeating: Boolean = false

    private val executor: ScheduledExecutorService = Executors.newSingleThreadScheduledExecutor()
    private var disconnectTask: ScheduledFuture<*>? = null

    fun queue(track: AudioTrack) {
        log.info("Adding {} in the queue", track.info.title)
        cancelDisconnectTimer()
        if (player.startTrack(track, true)) {
            logTrackStarted(track)
        } else {
            if (!queue.offer(track)) throw RuntimeException("Item was not added to the queue")
        }
    }

    fun nextTrack() {
        val track = queue.poll()
        if (track != null) {
            logTrackStarted(track)
        }
        player.startTrack(track, false)
    }

    fun skipTrack(position: Int) {
        if (position < 1) return

        if (position == 1) {
            nextTrack()
            return
        }

        val tracks = queue.toMutableList()
        if (position <= tracks.size) {
            tracks.removeAt(position - 1)
            queue.clear()
            queue.addAll(tracks)
        }
    }

    fun clear() {
        queue.clear()
        nextTrack()
        checkDisconnect()
    }

    fun getQueueSize(): Int = queue.size

    fun shuffle() {
        val tracks = queue.toMutableList()
        queue.clear()
        tracks.shuffle()
        tracks.forEach { queue(it) }
    }

    override fun onTrackEnd(player: AudioPlayer, track: AudioTrack, endReason: AudioTrackEndReason) {
        if (endReason.mayStartNext) {
            if (isRepeating) {
                player.startTrack(track.makeClone(), false)
                checkDisconnect()
                return
            }
            nextTrack()
            checkDisconnect()
            return
        }

        if (endReason == AudioTrackEndReason.LOAD_FAILED) {
            log.error("Error playing {} ({})", track.info.title, track.info.uri)
            nextTrack()
            checkDisconnect()
        }

        checkDisconnect()
    }

    override fun onTrackStart(player: AudioPlayer, track: AudioTrack) {
        cancelDisconnectTimer()
    }

    private fun checkDisconnect() {
        if (queue.isEmpty() && player.playingTrack == null) {
            scheduleDisconnect()
        }
    }

    private fun scheduleDisconnect() {
        disconnectTask?.cancel(false)
        disconnectTask = executor.schedule({
            guild.audioManager.closeAudioConnection()
        }, 3, TimeUnit.MINUTES)
    }

    private fun cancelDisconnectTimer() {
        disconnectTask?.cancel(false)
        disconnectTask = null
    }

    private fun logTrackStarted(track: AudioTrack) {
        val member = track.getUserData(Member::class.java)

        if (member != null) {
            log.info("Playing {} in {} requested by {}", track.info.title, member.guild.name, member.effectiveName)
        } else {
            log.info("Playing {}", track.info.title)
        }
    }
}
